lib/Sema/CodeSynthesis.cpp

```diff
@@ -1302,12 +1531,136 @@ void swift::maybeAddAccessorsToVariable(VarDecl *var, TypeChecker &TC) {
+    }
+    
+    {
+      // The type must refer to a protocol.
+      auto behaviorProtoTy = behaviorType->getAs<ProtocolType>();
+      if (!behaviorProtoTy) {
+        TC.diagnose(behavior->getLoc(),
+                    diag::property_behavior_not_protocol);
+        behavior->Conformance = (NormalProtocolConformance*)nullptr;
+        goto make_behavior_accessors;
+      }
+      auto behaviorProto = behaviorProtoTy->getDecl();
+
+      // Validate the behavior protocol and all its extensions so we can do
+      // name lookup.
+      TC.validateDecl(behaviorProto);
+      for (auto ext : behaviorProto->getExtensions()) {
+        TC.validateExtension(ext);
+      }
+      
+      // Look up the behavior protocol's "value" property, or bail if it doesn't
+      // have one. The property's accessors will decide whether the getter
+      // is mutating, and whether there's a setter. We'll type-check to make
+      // sure the property type matches later after validation.
+      auto lookup = TC.lookupMember(dc, behaviorProtoTy, TC.Context.Id_value);
+      for (auto found : lookup) {
+        if (auto foundVar = dyn_cast<VarDecl>(found.Decl)) {
+          if (valueProp) {
+            TC.diagnose(behavior->getLoc(),
+                        diag::property_behavior_protocol_reqt_ambiguous,
+                        TC.Context.Id_value);
+            TC.diagnose(valueProp->getLoc(),
+                        diag::property_behavior_protocol_reqt_here,
+                        TC.Context.Id_value);
+            TC.diagnose(foundVar->getLoc(),
+                        diag::property_behavior_protocol_reqt_here,
+                        TC.Context.Id_value);
+            break;
+          }
+            
+          valueProp = foundVar;
+        }
+      }
+      
+      if (!valueProp) {
+        TC.diagnose(behavior->getLoc(),
+                    diag::property_behavior_protocol_no_value);
+        goto make_behavior_accessors;
+      }
+      
+      TC.validateDecl(valueProp);
+      
+      // Set up a conformance to represent the behavior instantiation.
+      // The conformance will be on the containing 'self' type, or '()' if the
+      // property is in a non-type context.
+      Type behaviorSelf;
+      if (dc->isTypeContext()) {
+        behaviorSelf = dc->getSelfTypeInContext();
+        assert(behaviorSelf && "type context doesn't have self type?!");
+        if (var->isStatic())
+          behaviorSelf = MetatypeType::get(behaviorSelf);
+      } else {
+        behaviorSelf = TC.Context.TheEmptyTupleType;
+      }
+      
+      conformance = TC.Context.getBehaviorConformance(behaviorSelf,
+                                                       behaviorProto,
+                                                       behavior->getLoc(), dc,
+                                            ProtocolConformanceState::Checking);
+    }
+  make_behavior_accessors:
+    FuncDecl *getter = nullptr, *setter = nullptr;
```

Comment:
The only free variables in the block after the label are 'var', 'valueProp', 'conformance' and 'TC'. If you move this into its own function you won't need the gotos. This would move the var->setIsBeingTypeCheckedCalls() calls closer together which is good IMHO